File: net/JetBrains.FormatRipper.Tests/src/FakePki.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using JetBrains.Annotations;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.X509;

namespace JetBrains.SignatureVerifier.Tests
{
  class FakePki
  {
    private static readonly DerObjectIdentifier RSA_ENCRYPTION = new("1.2.840.113549.1.1.1");
    private static readonly DerObjectIdentifier SHA1_WITH_RSA_SIGNATURE = new("1.2.840.113549.1.1.5");
    private readonly AsymmetricCipherKeyPair _keyPair;
    private readonly AlgorithmIdentifier _signatureAlg = new(SHA1_WITH_RSA_SIGNATURE);
    private const int publicKeyLength = 1024;

    public X509Certificate Certificate { get; }
    public X509Crl Crl { get; private set; }
    private readonly List<X509Certificate> _certificates = new();
    public IReadOnlyCollection<X509Certificate> IssuedCertificates => _certificates.AsReadOnly();
    private Dictionary<Org.BouncyCastle.Math.BigInteger, DateTime> RevokedCertificates { get; } = new();

    public static FakePki CreateRoot([NotNull] string name, DateTime utcValidFrom, DateTime utcValidTo)
    {
      if (name == null) throw new ArgumentNullException(nameof(name));

      if (utcValidFrom >= utcValidTo)
        throw new ArgumentException($"{nameof(utcValidTo)} must be greater then {nameof(utcValidFrom)}");

      return new FakePki(name, utcValidFrom, utcValidTo);
    }

    private FakePki(string name, DateTime validFrom, DateTime validTo)
    {
      _keyPair = getNewPair();
      X509Name subject = new X509Name($"CN={name}");
      var res = enroll(subject, _keyPair, name, validFrom, validTo, 0, false, false);
      Certificate = res;
      Crl = createCrl();
    }


    public (AsymmetricCipherKeyPair keyPair, X509Certificate certificate) Enroll(string name, DateTime validFrom,
      DateTime validTo, bool codeSign)
    {
      var keyPair = getNewPair();
      var certificate = enroll(Certificate.SubjectDN, keyPair, name, validFrom, validTo, _certificates.Count + 1, true,
        codeSign);
      _certificates.Add(certificate);
      return (keyPair, certificate);
    }

    public void Revoke([NotNull] X509Certificate certificate, bool renewCrl)
    {
      if (certificate == null) throw new ArgumentNullException(nameof(certificate));

      if (isIssued(certificate))
      {
        RevokedCertificates.Add(certificate.SerialNumber, DateTime.UtcNow);

        if (renewCrl)
        {
          Crl = createCrl();
        }
      }
    }

    public void UpdateCrl()
    {
      Crl = createCrl();
    }

    private bool isIssued(X509Certificate certificate)
    {
      return certificate.IssuerDN.Equivalent(Certificate.SubjectDN);
    }

    private X509Certificate enroll(X509Name issuerDN, AsymmetricCipherKeyPair keyPair,
      string subjectName, DateTime validFrom,
      DateTime validTo,
      int sn,
      bool addCrlDp,
      bool codeSign)
    {
      var version = new DerTaggedObject(0, new DerInteger(2));
      var serialNumber = new DerInteger(sn);
      var startDate = new Time(validFrom);
      var endDate = new Time(validTo);
      var dates = new DerSequence(startDate, endDate);
      var subject = new X509Name($"CN={subjectName}");
      var alg = new AlgorithmIdentifier(RSA_ENCRYPTION);
      var rsaKeyParameters = keyPair.Public as RsaKeyParameters;
      Debug.Assert(rsaKeyParameters != null, nameof(rsaKeyParameters) + " != null");
      var keyData = new RsaPublicKeyStructure(rsaKeyParameters.Modulus, rsaKeyParameters.Exponent).GetEncoded();
      var subjectPublicKeyInfo = new SubjectPublicKeyInfo(alg, keyData);
      var vec = new Asn1EncodableVector(version, serialNumber, _signatureAlg, issuerDN, dates, subject,
        subjectPublicKeyInfo);

      var extOids = new List<DerObjectIdentifier>();
      var extValues = new List<X509Extension>();

      if (addCrlDp)
      {
        var names = new GeneralNames(new GeneralName(GeneralName.UniformResourceIdentifier,
          new DerIA5String("http://fakepki/crl")));

        var crlDistPoint = new CrlDistPoint(
          new[]
          {
            new DistributionPoint(new DistributionPointName(DistributionPointName.FullName, names), null, null)
          });

        extOids.Add(X509Extensions.CrlDistributionPoints);
        extValues.Add(new(false, new DerOctetString(crlDistPoint)));
      }

      if (codeSign)
      {
        extOids.Add(X509Extensions.ExtendedKeyUsage);
        extValues.Add(new(false, new DerOctetString(new DerSequence(KeyPurposeID.IdKPCodeSigning))));
      }

      Debug.Assert(extOids.Count == extValues.Count);

      if (extOids.Any())
      {
        var ext = new X509Extensions(extOids, extValues);
        vec.AddOptionalTagged(true, 3, ext);
      }

      var seq = DerSequence.FromVector(vec);
      var tbs = TbsCertificateStructure.GetInstance(seq);
      var tbsData = tbs.GetEncoded();
      var sig = sign(tbsData, _keyPair.Private);
      var cs = new X509CertificateStructure(tbs, _signatureAlg, new DerBitString(sig));

      return new X509Certificate(cs);
    }

    private X509Crl createCrl()
    {
      var version = new DerInteger(1);
      var issuer = Certificate.SubjectDN;
      var now = DateTime.UtcNow.AddMinutes(1);
      var thisUpdate = new Time(now);
      var nextUpdate = new Time(now.AddDays(5));
      var revokedCertificates = getRevokedCertificates();
      var seq = new DerSequence(version, _signatureAlg, issuer, thisUpdate, nextUpdate, revokedCertificates);
      var tbs = TbsCertificateList.GetInstance(seq);
      var tbsData = tbs.GetEncoded();
      var sig = sign(tbsData, _keyPair.Private);
      var certList = CertificateList.GetInstance(new DerSequence(tbs, _signatureAlg, new DerBitString(sig)));
      return new X509Crl(certList);
    }

    private DerSequence getRevokedCertificates()
    {
      var vec = Asn1EncodableVector.FromEnumerable(
        RevokedCertificates.Select(s => getRevokedCertificate(s.Key, s.Value)));
      return new DerSequence(vec);
    }

    private DerSequence getRevokedCertificate(Org.BouncyCastle.Math.BigInteger serialNumber, DateTime revocationTime)
    {
      return new DerSequence(new DerInteger(serialNumber), new Time(revocationTime));
    }

    private byte[] sign(byte[] data, AsymmetricKeyParameter key)
    {
      var signer = SignerUtilities.GetSigner(_signatureAlg.Algorithm.Id);
      signer.Init(true, key);
      signer.BlockUpdate(data, 0, data.Length);
      return signer.GenerateSignature();
    }

    private static AsymmetricCipherKeyPair getNewPair()
    {
      var keyPairGenerator = GeneratorUtilities.GetKeyPairGenerator(RSA_ENCRYPTION);
      keyPairGenerator.Init(new KeyGenerationParameters(SecureRandom.GetInstance("SHA1PRNG", true), publicKeyLength));
      var keyPair = keyPairGenerator.GenerateKeyPair();
      return keyPair;
    }
  }
}

File: jvm/src/test/kotlin/com/jetbrains/signatureverifier/tests/FakePki.kt
package com.jetbrains.signatureverifier.tests

import com.jetbrains.signatureverifier.crypt.Utils.ConvertToDate
import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.pkcs.RSAPublicKey
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x509.*
import org.bouncycastle.cert.X509CRLHolder
import org.bouncycastle.cert.X509CertificateHolder
import org.jetbrains.annotations.NotNull
import java.math.BigInteger
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.SecureRandom
import java.security.Signature
import java.time.Clock
import java.time.LocalDateTime
import java.util.*

open class FakePki {
  companion object {
    private val RSA_ENCRYPTION = ASN1ObjectIdentifier("1.2.840.113549.1.1.1")
    private val SHA1_WITH_RSA_SIGNATURE = ASN1ObjectIdentifier("1.2.840.113549.1.1.5")
    private const val publicKeyLength = 1024

    fun CreateRoot(@NotNull name: String, utcValidFrom: Date, utcValidTo: Date): FakePki {
      if (utcValidFrom >= utcValidTo)
        throw IllegalArgumentException("utcValidTo must be greater then utcValidFrom")

      return FakePki(name, utcValidFrom, utcValidTo)
    }

    private fun getNewPair(): KeyPair {
      val keyGen = KeyPairGenerator.getInstance("RSA")
      keyGen.initialize(publicKeyLength, SecureRandom.getInstance("SHA1PRNG"))
      return keyGen.generateKeyPair()
    }
  }

  private val _keyPair: KeyPair = getNewPair()
  private val _signatureAlg = AlgorithmIdentifier(SHA1_WITH_RSA_SIGNATURE)
  private var _certificate: X509CertificateHolder

  val Certificate: X509CertificateHolder
    get() = _certificate

  var _crl: X509CRLHolder? = null

  var Crl: X509CRLHolder?
    get() = _crl
    set(value) {
      _crl = value
    }

  private val _certificates = mutableListOf<X509CertificateHolder>()

  val IssuedCertificates: Collection<X509CertificateHolder>
    get() = _certificates

  private val RevokedCertificates = mutableMapOf<BigInteger, Date>()

  constructor(name: String, validFrom: Date, validTo: Date) {
    val subject = X500Name("CN=${name}")
    val res = enroll(subject, _keyPair, name, validFrom, validTo, 0, false, false)
    _certificate = res
    Crl = createCrl()
  }

  fun Enroll(name: String, validFrom: Date, validTo: Date, codeSign: Boolean): Pair<KeyPair, X509CertificateHolder> {
    val keyPair = getNewPair()
    val certificate = enroll(Certificate.subject, keyPair, name, validFrom, validTo, (_certificates.count() + 1).toLong(), true, codeSign)
    _certificates.add(certificate)
    return Pair(keyPair, certificate)
  }

  fun Revoke(@NotNull certificate: X509CertificateHolder, renewCrl: Boolean) {
    if (isIssued(certificate)) {
      RevokedCertificates.put(certificate.serialNumber, LocalDateTime.now().ConvertToDate())
      if (renewCrl) {
        Crl = createCrl()
      }
    }
  }

  fun UpdateCrl() {
    Crl = createCrl()
  }

  private fun isIssued(certificate: X509CertificateHolder): Boolean {
    return certificate.issuer.equals(Certificate.subject)
  }

  private fun enroll(issuerDN: X500Name, keyPair: KeyPair, subjectName: String, validFrom: Date, validTo: Date, sn: Long, addCrlDp: Boolean, codeSign: Boolean): X509CertificateHolder {
    val version = DERTaggedObject(0, ASN1Integer(2))
    val serialNumber = ASN1Integer(sn)
    val startDate = Time(validFrom)
    val endDate = Time(validTo)
    val dates = DERSequence(arrayOf(startDate, endDate))
    val subject = X500Name("CN=${subjectName}")
    val alg = AlgorithmIdentifier(RSA_ENCRYPTION)
    val rsaKeyParameters = keyPair.public as java.security.interfaces.RSAPublicKey
    val keyData = RSAPublicKey(rsaKeyParameters.modulus, rsaKeyParameters.publicExponent).encoded
    val subjectPublicKeyInfo = SubjectPublicKeyInfo(alg, keyData)
    val vec = ASN1EncodableVector().also { it.addAll(arrayOf(version, serialNumber, _signatureAlg, issuerDN, dates, subject, subjectPublicKeyInfo)) }
    val extValues = mutableListOf<Extension>()

    if (addCrlDp) {
      val names = GeneralNames(GeneralName(GeneralName.uniformResourceIdentifier, DERIA5String("http://fakepki/crl")))
      val crlDistPoint = CRLDistPoint(arrayOf(DistributionPoint(DistributionPointName(DistributionPointName.FULL_NAME, names), null, null)))
      extValues.add(Extension.create(Extension.cRLDistributionPoints, false, crlDistPoint))
    }

    if (codeSign) {
      extValues.add(Extension.create(Extension.extendedKeyUsage, false, DERSequence(KeyPurposeId.id_kp_codeSigning)))
    }

    if (extValues.any()) {
      val ext = Extensions(extValues.toTypedArray())
      vec.addOptionalTagged(true, 3, ext)
    }

    val seq = DERSequence(vec)
    val tbs = TBSCertificate.getInstance(seq)
    val tbsData = tbs.encoded
    val sig = sign(tbsData, _keyPair)
    val cs =
      org.bouncycastle.asn1.x509.Certificate.getInstance(getDerSequenceInstance(tbs, _signatureAlg, DERBitString(sig)))
    return X509CertificateHolder(cs)
  }

  private fun createCrl(): X509CRLHolder {
    val version = ASN1Integer(1)
    val issuer = Certificate.subject
    val now = LocalDateTime.now(Clock.systemUTC()).plusMinutes(1)
    val thisUpdate = Time(now.ConvertToDate())
    val nextUpdate = Time(now.plusDays(5).ConvertToDate())
    val revokedCertificates = getRevokedCertificates()
    val seq = getDerSequenceInstance(version, _signatureAlg, issuer, thisUpdate, nextUpdate, revokedCertificates)
    val tbs = TBSCertList.getInstance(seq)
    val tbsData = tbs.encoded
    val sig = sign(tbsData, _keyPair)
    val certList = CertificateList.getInstance(getDerSequenceInstance(tbs, _signatureAlg, DERBitString(sig)))
    return X509CRLHolder(certList)
  }

  private fun getRevokedCertificates(): DERSequence {
    val vec = RevokedCertificates.map { s -> getRevokedCertificate(s.key, s.value) }.toAsn1EncodableVector()
    return DERSequence(vec)
  }

  private fun getRevokedCertificate(serialNumber: BigInteger, revocationTime: Date): DERSequence {
    return DERSequence(arrayOf(ASN1Integer(serialNumber), Time(revocationTime)))
  }

  private fun sign(data: ByteArray, key: KeyPair): ByteArray {
    val signature = Signature.getInstance(_signatureAlg.algorithm.id).also { it.initSign(key.private) }
    signature.update(data)
    return signature.sign()
  }

  private fun ASN1EncodableVector.addOptionalTagged(isExplicit: Boolean, tagNo: Int, obj: ASN1Encodable?) {
    if (null != obj) {
      this.add(DERTaggedObject(isExplicit, tagNo, obj))
    }
  }

  private fun Collection<ASN1Encodable>.toAsn1EncodableVector(): ASN1EncodableVector {
    val v = ASN1EncodableVector()
    forEach { v.add(it) }
    return v
  }

  private fun getDerSequenceInstance(vararg asn1Encodable: ASN1Encodable): DERSequence {
    return DERSequence(ASN1EncodableVector().also { it.addAll(asn1Encodable) })
  }
}
